File: CompInterpreter/server/src/interprete/operaciones.js
/* ============================================================
   Operaciones aritmeticas, relacionales y logicas.
   Las tablas de compatibilidad son LITERALES del enunciado (5.5).
   Propagacion por null: si un operando es JS null (ya hubo un error
   antes), se devuelve null en silencio (un error por causa raiz).
   OJO: Valor de tipo 'null' (el null DEL LENGUAJE) NO es JS null.
   ============================================================ */
const { TIPO } = require('./tipos');
const { Valor, aNumero, aTexto } = require('./valor');

// I=int, D=double, B=bool, C=char, S=string ; 'X' = combinacion invalida
// Cada tabla: filas = operando izquierdo, columnas = operando derecho.

// 5.5.1 Suma
const T_SUMA = {
  int:    { int: 'int',    double: 'double', bool: 'int', char: 'int',  string: 'string' },
  double: { int: 'double', double: 'double', bool: 'double', char: 'double', string: 'string' },
  bool:   { int: 'int',    double: 'double', bool: 'X',   char: 'X',   string: 'string' },
  char:   { int: 'int',    double: 'double', bool: 'X',   char: 'string', string: 'string' },
  string: { int: 'string', double: 'string', bool: 'string', char: 'string', string: 'string' }
};
// 5.5.2 Resta
const T_RESTA = {
  int:    { int: 'int',    double: 'double', bool: 'int', char: 'int' },
  double: { int: 'double', double: 'double', bool: 'double', char: 'double' },
  bool:   { int: 'int',    double: 'double', bool: 'X',   char: 'X' },
  char:   { int: 'int',    double: 'double', bool: 'X',   char: 'X' }
};
// 5.5.3 Multiplicacion
const T_MULT = {
  int:    { int: 'int',    double: 'double', char: 'int' },
  double: { int: 'double', double: 'double', char: 'double' },
  char:   { int: 'int',    double: 'double', char: 'X' }
};
// 5.5.4 Division (siempre Decimal donde esta definida)
const T_DIV = {
  int:    { int: 'double', double: 'double', char: 'double' },
  double: { int: 'double', double: 'double', char: 'double' },
  char:   { int: 'double', double: 'double', char: 'X' }
};
// 5.5.5 Potencia
const T_POT = {
  int:    { int: 'int',    double: 'double' },
  double: { int: 'double', double: 'double' }
};
// 5.5.6 Raiz
const T_RAIZ = {
  int:    { int: 'double', double: 'double' },
  double: { int: 'double', double: 'double' }
};
// 5.5.6 Modulo  -> NOTA: el enunciado define Entero % Entero = DECIMAL
// (identico a la tabla de Raiz). Es inusual (uno esperaria Entero), pero
// se implementa TAL CUAL lo dice la fuente, sin "corregirlo".
const T_MOD = {
  int:    { int: 'double', double: 'double' },
  double: { int: 'double', double: 'double' }
};

const NOMBRE_OP = { '+': 'suma', '-': 'resta', '*': 'multiplicación',
  '/': 'división', '^': 'potencia', '$': 'raíz', '%': 'módulo' };

function nombreTipo(t) {
  return { int: 'ENTERO', double: 'DECIMAL', bool: 'BOOLEAN',
    char: 'CARÁCTER', string: 'CADENA', null: 'NULO', vector: 'VECTOR' }[t] || t;
}

function errBinaria(errores, op, ti, td, l, c) {
  errores.semantico('No se puede realizar la ' + NOMBRE_OP[op] + ' entre ' +
    nombreTipo(ti) + ' y ' + nombreTipo(td), l, c);
  return null;
}

// Empaqueta el numero como int o double segun el tipo resultante de la tabla.
function empacar(tipoRes, numero) {
  if (tipoRes === TIPO.INT) return Valor.int(numero);
  return Valor.double(numero);
}

function aritmetica(op, izq, der, l, c, errores) {
  if (izq === null || der === null) return null;   // propagacion por null
  let tabla;
  switch (op) {
    case '+': tabla = T_SUMA; break;
    case '-': tabla = T_RESTA; break;
    case '*': tabla = T_MULT; break;
    case '/': tabla = T_DIV; break;
    case '^': tabla = T_POT; break;
    case '$': tabla = T_RAIZ; break;
    case '%': tabla = T_MOD; break;
  }
  const fila = tabla[izq.tipo];
  const tipoRes = fila ? fila[der.tipo] : undefined;
  if (!tipoRes || tipoRes === 'X') return errBinaria(errores, op, izq.tipo, der.tipo, l, c);

  // caso string -> concatenacion
  if (tipoRes === TIPO.STRING) {
    return Valor.string(aTexto(izq) + aTexto(der));
  }
  const a = aNumero(izq), b = aNumero(der);
  let r;
  switch (op) {
    case '+': r = a + b; break;
    case '-': r = a - b; break;
    case '*': r = a * b; break;
    case '/':
      if (b === 0) { errores.semantico('División entre cero', l, c); return null; }
      r = a / b; break;
    case '^': r = Math.pow(a, b); break;
    case '$':
      if (b === 0) { errores.semantico('Índice de raíz igual a cero', l, c); return null; }
      r = Math.pow(a, 1 / b); break;
    case '%':
      if (b === 0) { errores.semantico('Módulo entre cero', l, c); return null; }
      r = a % b; break;
  }
  if (tipoRes === TIPO.INT) r = Math.trunc(r);
  return empacar(tipoRes, r);
}

// 5.5.7 Negacion unaria: solo int/double
function negacion(v, l, c, errores) {
  if (v === null) return null;
  if (v.tipo === TIPO.INT) return Valor.int(-v.valor);
  if (v.tipo === TIPO.DOUBLE) return Valor.double(-v.valor);
  errores.semantico('No se puede aplicar la negación unaria sobre ' + nombreTipo(v.tipo), l, c);
  return null;
}

/* ---------- Relacionales ----------
   DECISION DE DISENO (la matriz 5.6 del enunciado quedo con las celdas
   vacias al convertir el PDF; no es recuperable). Criterio adoptado y
   documentado aqui como decision propia, NO como dato del enunciado:
     - Igualdad (==, !=):
         * Nulo contra CUALQUIER tipo (solo verifica igualdad de nulidad).
         * int/double/char entre si (por valor numerico / codigo ASCII).
         * string SOLO con string.
         * bool SOLO con bool.
         * cualquier otra combinacion -> error semantico.
     - Orden (<, <=, >, >=):
         * int/double/char entre si (por valor / ASCII).
         * string con string (orden lexicografico).
         * cualquier otra combinacion (incluye bool y null) -> error semantico.
*/
function esNumComparable(t) { return t === TIPO.INT || t === TIPO.DOUBLE || t === TIPO.CHAR; }

function relacional(op, izq, der, l, c, errores) {
  if (izq === null || der === null) return null;
  const ti = izq.tipo, td = der.tipo;

  if (op === '==' || op === '!=') {
    let iguales;
    if (ti === TIPO.NULL || td === TIPO.NULL) {
      iguales = (ti === TIPO.NULL && td === TIPO.NULL);
    } else if (esNumComparable(ti) && esNumComparable(td)) {
      iguales = aNumero(izq) === aNumero(der);
    } else if (ti === TIPO.STRING && td === TIPO.STRING) {
      iguales = izq.valor === der.valor;
    } else if (ti === TIPO.BOOL && td === TIPO.BOOL) {
      iguales = izq.valor === der.valor;
    } else {
      errores.semantico('No se puede comparar (' + op + ') entre ' +
        nombreTipo(ti) + ' y ' + nombreTipo(td), l, c);
      return null;
    }
    return Valor.bool(op === '==' ? iguales : !iguales);
  }

  // operadores de orden
  let a, b;
  if (esNumComparable(ti) && esNumComparable(td)) {
    a = aNumero(izq); b = aNumero(der);
  } else if (ti === TIPO.STRING && td === TIPO.STRING) {
    a = izq.valor; b = der.valor;   // comparacion lexicografica de JS
  } else {
    errores.semantico('No se puede comparar (' + op + ') entre ' +
      nombreTipo(ti) + ' y ' + nombreTipo(td), l, c);
    return null;
  }
  let res;
  switch (op) {
    case '<': res = a < b; break;
    case '<=': res = a <= b; break;
    case '>': res = a > b; break;
    case '>=': res = a >= b; break;
  }
  return Valor.bool(res);
}

function logica(op, izq, der, l, c, errores) {
  if (op === '!') {
    if (izq === null) return null;
    if (izq.tipo !== TIPO.BOOL) {
      errores.semantico('El operador ! requiere un valor BOOLEAN, no ' + nombreTipo(izq.tipo), l, c);
      return null;
    }
    return Valor.bool(!izq.valor);
  }
  if (izq === null || der === null) return null;
  if (izq.tipo !== TIPO.BOOL || der.tipo !== TIPO.BOOL) {
    errores.semantico('El operador ' + op + ' requiere valores BOOLEAN, no ' +
      nombreTipo(izq.tipo) + ' y ' + nombreTipo(der.tipo), l, c);
    return null;
  }
  const r = op === '&&' ? (izq.valor && der.valor) : (izq.valor || der.valor);
  return Valor.bool(r);
}

module.exports = { aritmetica, negacion, relacional, logica, nombreTipo };

File: CompScript/src/main/java/compscript/interprete/Operaciones.java
package compscript.interprete;

import compscript.interprete.Tipo.Cat;

/**
 * Semantica de los operadores (5.5, 5.6, 5.7) y de los casteos (5.13).
 * Implementa LITERALMENTE las 8 tablas de compatibilidad aritmetica y la
 * matriz relacional del enunciado: toda combinacion no listada es un error
 * de tipo semantico que ABORTA la ejecucion (via Entorno.errorSemantico).
 */
public class Operaciones {

    /* ================= aritmetica (operadores infijos) ================= */

    public static Valor aritmetica(String op, Valor a, Valor b, Entorno e, int l, int c) {
        switch (op) {
            case "+": return suma(a, b, e, l, c);
            case "-": return restaMult(op, a, b, e, l, c);
            case "*": return restaMult(op, a, b, e, l, c);
            case "/": return division(a, b, e, l, c);
            case "^": return potencia(a, b, e, l, c);
            case "$": return raiz(a, b, e, l, c);
            case "%": return modulo(a, b, e, l, c);
            default:  return null;
        }
    }

    /* ---- 5.5.1 Suma (unica que admite bool, char y cadena) ---- */
    private static Valor suma(Valor a, Valor b, Entorno e, int l, int c) {
        Cat x = a.tipo.cat, y = b.tipo.cat;
        Cat res = null;
        if (x == Cat.STRING || y == Cat.STRING) {
            // Cadena con cualquiera de los otros primitivos -> Cadena
            if (esPrimitivo(x) && esPrimitivo(y)) res = Cat.STRING;
        } else if (x == Cat.CHAR && y == Cat.CHAR) {
            res = Cat.STRING;                       // char + char -> cadena
        } else if (x == Cat.INT && y == Cat.INT) {
            res = Cat.INT;
        } else if ((x == Cat.INT || x == Cat.DOUBLE || x == Cat.BOOL || x == Cat.CHAR)
                && (y == Cat.INT || y == Cat.DOUBLE || y == Cat.BOOL || y == Cat.CHAR)) {
            // numerico/bool/char: si alguno es double -> Decimal, si no -> Entero
            boolean hayDouble = (x == Cat.DOUBLE || y == Cat.DOUBLE);
            boolean boolConBool = (x == Cat.BOOL && y == Cat.BOOL);
            boolean boolConChar = (x == Cat.BOOL && y == Cat.CHAR) || (x == Cat.CHAR && y == Cat.BOOL);
            if (boolConBool || boolConChar) res = null;   // invalidas en la tabla
            else res = hayDouble ? Cat.DOUBLE : Cat.INT;
        }
        if (res == null) return error("suma", a, b, e, l, c);
        if (res == Cat.STRING) return Valor.vString(concat(a) + concat(b));
        double r = a.numero() + b.numero();
        return res == Cat.INT ? Valor.vInt((int) r) : Valor.vDouble(r);
    }

    /* ---- 5.5.2 Resta / 5.5.3 Multiplicacion (Entero, Decimal, Caracter) ---- */
    private static Valor restaMult(String op, Valor a, Valor b, Entorno e, int l, int c) {
        Cat res = tablaRestaMult(a.tipo.cat, b.tipo.cat);
        if (res == null) return error(op.equals("-") ? "resta" : "multiplicacion", a, b, e, l, c);
        double r = op.equals("-") ? a.numero() - b.numero() : a.numero() * b.numero();
        return res == Cat.INT ? Valor.vInt((int) r) : Valor.vDouble(r);
    }

    private static Cat tablaRestaMult(Cat x, Cat y) {
        if (!enGrupo(x, Cat.INT, Cat.DOUBLE, Cat.CHAR) || !enGrupo(y, Cat.INT, Cat.DOUBLE, Cat.CHAR))
            return null;
        if (x == Cat.CHAR && y == Cat.CHAR) return null;       // char - char invalido
        return (x == Cat.DOUBLE || y == Cat.DOUBLE) ? Cat.DOUBLE : Cat.INT;
    }

    /* ---- 5.5.4 Division (siempre Decimal) ---- */
    private static Valor division(Valor a, Valor b, Entorno e, int l, int c) {
        Cat x = a.tipo.cat, y = b.tipo.cat;
        if (!enGrupo(x, Cat.INT, Cat.DOUBLE, Cat.CHAR) || !enGrupo(y, Cat.INT, Cat.DOUBLE, Cat.CHAR)
                || (x == Cat.CHAR && y == Cat.CHAR))
            return error("division", a, b, e, l, c);
        if (b.numero() == 0) e.errorSemantico("division entre cero", l, c);
        return Valor.vDouble(a.numero() / b.numero());
    }

    /* ---- 5.5.5 Potencia (Entero/Decimal) ---- */
    private static Valor potencia(Valor a, Valor b, Entorno e, int l, int c) {
        Cat x = a.tipo.cat, y = b.tipo.cat;
        if (!enGrupo(x, Cat.INT, Cat.DOUBLE) || !enGrupo(y, Cat.INT, Cat.DOUBLE))
            return error("potencia", a, b, e, l, c);
        double r = Math.pow(a.numero(), b.numero());
        return (x == Cat.INT && y == Cat.INT) ? Valor.vInt((int) r) : Valor.vDouble(r);
    }

    /* ---- 5.5.6 Raiz (a $ b = raiz b-esima de a; siempre Decimal) ---- */
    private static Valor raiz(Valor a, Valor b, Entorno e, int l, int c) {
        Cat x = a.tipo.cat, y = b.tipo.cat;
        if (!enGrupo(x, Cat.INT, Cat.DOUBLE) || !enGrupo(y, Cat.INT, Cat.DOUBLE))
            return error("raiz", a, b, e, l, c);
        if (b.numero() == 0) e.errorSemantico("raiz de indice cero", l, c);
        return Valor.vDouble(Math.pow(a.numero(), 1.0 / b.numero()));
    }

    /* ---- 5.5.7 Modulo (Entero/Decimal; resultado Decimal) ---- */
    private static Valor modulo(Valor a, Valor b, Entorno e, int l, int c) {
        Cat x = a.tipo.cat, y = b.tipo.cat;
        if (!enGrupo(x, Cat.INT, Cat.DOUBLE) || !enGrupo(y, Cat.INT, Cat.DOUBLE))
            return error("modulo", a, b, e, l, c);
        if (b.numero() == 0) e.errorSemantico("modulo entre cero", l, c);
        return Valor.vDouble(a.numero() % b.numero());
    }

    /* ---- 5.5.8 Negacion unaria (Entero/Decimal) ---- */
    public static Valor negacion(Valor a, Entorno e, int l, int c) {
        if (a.tipo.cat == Cat.INT) return Valor.vInt(-(Integer) a.valor);
        if (a.tipo.cat == Cat.DOUBLE) return Valor.vDouble(-(Double) a.valor);
        e.errorSemantico("no se puede negar un valor de tipo " + a.tipo.nombre().toUpperCase(), l, c);
        return null;
    }

    /* ================= relacionales (5.6) ================= */

    public static Valor relacional(String op, Valor a, Valor b, Entorno e, int l, int c) {
        Cat x = a.tipo.cat, y = b.tipo.cat;
        boolean ok = (esNum(x) && esNum(y))            // Entero/Decimal/Caracter entre si
                || (x == Cat.BOOL && y == Cat.BOOL)     // Booleano solo con Booleano
                || (x == Cat.STRING && y == Cat.STRING); // Cadena solo con Cadena
        if (!ok) return error("comparacion (" + op + ")", a, b, e, l, c);

        int cmp;
        if (x == Cat.STRING) cmp = ((String) a.valor).compareTo((String) b.valor);
        else cmp = Double.compare(a.numero(), b.numero());

        boolean r;
        switch (op) {
            case "==": r = cmp == 0; break;
            case "!=": r = cmp != 0; break;
            case "<":  r = cmp < 0;  break;
            case "<=": r = cmp <= 0; break;
            case ">":  r = cmp > 0;  break;
            case ">=": r = cmp >= 0; break;
            default:   r = false;
        }
        return Valor.vBool(r);
    }

    /* ================= logicos (5.7) ================= */

    public static Valor logica(String op, Valor a, Valor b, Entorno e, int l, int c) {
        if (a.tipo.cat != Cat.BOOL || b.tipo.cat != Cat.BOOL)
            return error(op.equals("&&") ? "AND (&&)" : "OR (||)", a, b, e, l, c);
        boolean ba = (Boolean) a.valor, bb = (Boolean) b.valor;
        return Valor.vBool(op.equals("&&") ? (ba && bb) : (ba || bb));
    }

    public static Valor negacionLogica(Valor a, Entorno e, int l, int c) {
        if (a.tipo.cat != Cat.BOOL)
            e.errorSemantico("el operador ! requiere un Booleano, no " + a.tipo.nombre().toUpperCase(), l, c);
        return Valor.vBool(!(Boolean) a.valor);
    }

    /* ================= casteos (5.13) ================= */

    public static Valor cast(Valor v, Tipo destino, Entorno e, int l, int c) {
        Cat o = v.tipo.cat, d = destino.cat;
        if (o == d) return new Valor(destino, v.valor);          // identidad
        if (o == Cat.INT  && d == Cat.DOUBLE) return Valor.vDouble((Integer) v.valor);
        if (o == Cat.DOUBLE && d == Cat.INT)  return Valor.vInt((int) (double) (Double) v.valor);
        if (o == Cat.INT  && d == Cat.CHAR)   return Valor.vChar((char) (int) (Integer) v.valor);
        if (o == Cat.CHAR && d == Cat.INT)    return Valor.vInt((Character) v.valor);
        if (o == Cat.CHAR && d == Cat.DOUBLE) return Valor.vDouble((int) (Character) v.valor);
        e.errorSemantico("no se puede castear " + v.tipo.nombre().toUpperCase()
                + " a " + destino.nombre().toUpperCase(), l, c);
        return null;
    }

    /* ================= helpers ================= */

    private static boolean esPrimitivo(Cat c) {
        return c == Cat.INT || c == Cat.DOUBLE || c == Cat.BOOL || c == Cat.CHAR || c == Cat.STRING;
    }
    private static boolean esNum(Cat c) {
        return c == Cat.INT || c == Cat.DOUBLE || c == Cat.CHAR;
    }
    private static boolean enGrupo(Cat c, Cat... permitidos) {
        for (Cat p : permitidos) if (c == p) return true;
        return false;
    }

    /** Representacion de un valor primitivo para la concatenacion de cadenas. */
    private static String concat(Valor v) {
        return v.texto();
    }

    private static Valor error(String operacion, Valor a, Valor b, Entorno e, int l, int c) {
        e.errorSemantico("No se puede realizar " + operacion + " entre "
                + a.tipo.nombre().toUpperCase() + " y " + b.tipo.nombre().toUpperCase(), l, c);
        return null;
    }
}
